homeassistant/components/strava/__init__.py

```diff
@@ -0,0 +1,288 @@
+        """Initialize strava data model."""
+        from stravalib.client import Client
+
+        self.client = Client()
+        self._configurator = None
+        self._token = None
+        self._hass = hass
+        self._config = config
+
+        self._hass.http.register_view(StravaAuthCallbackView(self))
+
+        self._client_id = config.get(CONF_CLIENT_ID)
+        self._client_secret = config.get(CONF_CLIENT_SECRET)
+
+        self.athletes = {}
+        self.gears = {}
+        self.clubs = {}
+
+    @property
+    def is_authorized(self):
+        """Check if there is a (possiblly expired) OAuth2 token."""
+        return self._token is not None
+
+    @property
+    def is_token_valid(self):
+        """Check if OAuth2 token is present and not expired."""
+        if not self.is_authorized:
+            _LOGGER.error("Not authorized")
+            return False
+
+        expires_at = datetime.fromtimestamp(self._token['expires_at'])
+        if expires_at > datetime.now() + 300:
+            return True
+
+        _LOGGER.info("Token expired: %s", repr(self._token))
+        return False
+
+    async def get_token(self):
+        """Load the OAuth2 token from the store."""
+        if not self.is_authorized:
+            store = self._hass.helpers.storage.Store(STORAGE_VERSION,
+                                                     STORAGE_KEY)
+            self._token = await store.async_load()
+
+            if self._token:
+                self.client.access_token = self._token['access_token']
+            else:
+                _LOGGER.info("Requesting token")
+                await self.request_token()
+                return
+
+        expires_at = datetime.fromtimestamp(self._token['expires_at'])
+        if expires_at < datetime.now():
+            await self.refresh_token()
+
+    async def authorize(self, code, hass):
+        """Request initial authorization."""
+        self._token = await hass.async_add_executor_job(
+            self.client.exchange_code_for_token,
+            self._client_id,
+            self._client_secret,
+            code
+        )
+
+        store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+        await store.async_save(self._token)
+
+        if self.is_authorized:
+            await hass.async_add_executor_job(
+                hass.components.configurator.request_done,
+                self._configurator
+            )
+            del self._configurator
+
+        await async_setup(hass, self._config)
+
+    async def request_token(self):
+        """Request Strava access token."""
+        callback_url = '{}{}'.format(self._hass.config.api.base_url,
+                                     AUTH_CALLBACK_PATH)
+        authorize_url = self.client.authorization_url(
+            client_id=self._config.get(CONF_CLIENT_ID),
+            redirect_uri=callback_url)
+
+        self._configurator = \
+            self._hass.components.configurator.async_request_config(
```

Comment:
The configurator is deprecated. New integrations must use a config flow instead.

See https://developers.home-assistant.io/docs/en/config_entries_index.html

See the somfy integration for an example that uses oauth2 and a callback view in the config flow.